[iOS] "A problem repeatedly occurred" error page when safari resuming from background

cdumez opened the pull request:
#### e6e28097b53cb99caa8fbf8440f215643e0c0dba
<pre>
[iOS] &quot;A problem repeatedly occurred&quot; error page when safari resuming from background
<a href="https://bugs.webkit.org/show_bug.cgi?id=242292">https://bugs.webkit.org/show_bug.cgi?id=242292</a>
&lt;rdar://96178034&gt;

Reviewed by Darin Adler.

When Safari is suspended in the background, its WebProcesses are idle and thus more
likely to get jetsammed. If they get jetsammed while Safari was suspended, Safari
will only receive the crash notification(s) once it is resumed.

The issue was that when resuming Safari after several WebProcesses have been jetsammed,
we would get a crash notification for a view&apos;s process and let Safari know. This would
cause Safari to call reload on the view to recover from the crash. However, instead of
launching a fresh new process, WebPageProxy::launchProcess() would sometimes try and
be smart and reuse an eligible existing WebProcess, which is usually fine. However,
we can get unlucky and choose a WebProcess that was also jetsammed while suspended,
for which we haven&apos;t received the crash notification yet (but are about to).
When the crash notification for this second process comes in shortly after, we tell
Safari that the view&apos;s process crashes again. This trips some logic in Safari that
causes it to show the &quot;A problem repeatedly occurred&quot; error page since it was told
that the view&apos;s process crashed twice in a row (i.e. failed to recover).

To address the issue, I made the following change:
- WebPageProxy::launchProcess() now only reuses the related WKWebView&apos;s process
  when it is its initial process, not in case of a crash. This was the case in
  the radar.

This fix is covered by a new API test that reproduces the scenario in the radar.

Ideally, we&apos;d prevent WebPageProxy::launchProcess() from reusing processes from
the process cache or suspended pages if they were terminated too. I will look
into this in a follow-up.

* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::findReusableSuspendedPageProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::takeProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForRegistrableDomain):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebContentProcessDidTerminate.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/252108@main">https://commits.webkit.org/252108@main</a>
</pre>

darinadler reviewed (approved):
Does the test cover all 3 of the improvements? These seem like the kinds of things that are easy to mess up when making other changes in the future so testing all 3 seems valuable if possible.

cdumez committed:
[iOS] "A problem repeatedly occurred" error page when safari resuming from background
https://bugs.webkit.org/show_bug.cgi?id=242292
<rdar://96178034>

Reviewed by Darin Adler.

When Safari is suspended in the background, its WebProcesses are idle and thus more
likely to get jetsammed. If they get jetsammed while Safari was suspended, Safari
will only receive the crash notification(s) once it is resumed.

The issue was that when resuming Safari after several WebProcesses have been jetsammed,
we would get a crash notification for a view's process and let Safari know. This would
cause Safari to call reload on the view to recover from the crash. However, instead of
launching a fresh new process, WebPageProxy::launchProcess() would sometimes try and
be smart and reuse an eligible existing WebProcess, which is usually fine. However,
we can get unlucky and choose a WebProcess that was also jetsammed while suspended,
for which we haven't received the crash notification yet (but are about to).
When the crash notification for this second process comes in shortly after, we tell
Safari that the view's process crashes again. This trips some logic in Safari that
causes it to show the "A problem repeatedly occurred" error page since it was told
that the view's process crashed twice in a row (i.e. failed to recover).

To address the issue, I made the following change:
- WebPageProxy::launchProcess() now only reuses the related WKWebView's process
  when it is its initial process, not in case of a crash. This was the case in
  the radar.

This fix is covered by a new API test that reproduces the scenario in the radar.

Ideally, we'd prevent WebPageProxy::launchProcess() from reusing processes from
the process cache or suspended pages if they were terminated too. I will look
into this in a follow-up.

* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::findReusableSuspendedPageProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::takeProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForRegistrableDomain):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebContentProcessDidTerminate.mm:
(TEST):

Canonical link: https://commits.webkit.org/252108@main
